File: PythonCode/Repo_CompIntroQP_Python/ChapterTD/Exercise_5_5_2.py
"""
 This script calculates the probability to remain in the the 
 initial state for a particle in a harmonic osciallator which
 is shaken. It does so by first expressing the kinetic energy term
 via FFT and then approximating the second order Magnus propagator as 
 a split operator.
 
 The harmonic oscillator potential is subject to a time-dependent 
 translation which shifts it in both directions before restoring it 
 in its original position. In addition to time, it takes omega as an 
 input; the duraton of the time-dependent shift is related to omega by 
 T = 2*pi/omega.

 Inputs
 L         - size of domain 
 N         - number of grid points (should be 2^n)
 Nt        - the number of time steps for each run
 k         - the strength of the potential
 omegaMin  - minimal omega value
 omegaMax  - maximal omega value
 domega    - step size in omega
 InitialState - the index of the initial state;  InitialState = 0 is 
 the ground state

 All inputs are hard coded initially
"""

# Libraries
import numpy as np
from matplotlib import pyplot as plt
from scipy import linalg


# Grid parameters
L = 10
N = 512              

# Strength of the harmonic oscillator
Kpot = 1

# Time parameters
Nstep = 500
omegaMin = 0.05
omegaMax = 1.5
domega = 0.025

# Index of initial state (ground state corresponds to zero)
InitialState = 0

# Shape of the potential
def Vpot(x):
    return 0.5*Kpot*x**2

# The displacement of the potential
def Ftrans(t, omega): 
    return 8/3**(3/2)*np.sin(omega*t/2)**2*np.sin(omega*t)

# Set up grid
x = np.linspace(-L/2, L/2, N)
h = L/(N-1)

# Kinetic energy operator (FFT)
k_max = np.pi/h
dk = 2*k_max/N
k = np.append(np.linspace(0, k_max-dk, int(N/2)), 
              np.linspace(-k_max, -dk, int(N/2)))
# Transform identity matrix
Tmat = np.fft.fft(np.identity(N, dtype=complex), axis = 0)
# Multiply by (ik)^2
Tmat = np.matmul(np.diag(-k**2), Tmat)
# Transform back to x-representation. 
Tmat = np.fft.ifft(Tmat, axis = 0)
# Correct pre-factor (unit mass and \hbar = 1)
Tmat = -1/2*Tmat    

# Add potential
# Full Hamiltonian
Ham = Tmat + np.diag(Vpot(x))

# Diagaonalize time-independent Hamiltonian (Hermitian matrix)
Evector, Bmat = np.linalg.eigh(Ham)
# Normalize eigenstates
Bmat = Bmat/np.sqrt(h)

# InitialState
Psi0 = Bmat[:, InitialState]

# Loop over the omega values
omegaVector = np.arange(omegaMin, omegaMax+domega/10, domega)
Pvector = 0*omegaVector          # Allocate    
index = 0
for omega in omegaVector:
  # Duration and time-step
  tTotal = 2*np.pi/omega
  dt = tTotal/Nstep
  # Propagator for half-step with kinetic energy
  UkinHalf = linalg.expm(-1j*Tmat*dt/2)
  t = 0
  # Initial state
  Psi = Psi0
 
  # Do the dynamics
  while t < tTotal:
     # Half step with kinetic energy
     Psi = np.matmul(UkinHalf, Psi)  
     # Full step with poetntial energy
     Translation = Ftrans(t+dt/2, omega)
     UpotFull = np.diag(np.exp(-1j*Vpot(x-Translation)*dt))
     Psi = np.matmul(UpotFull, Psi)
     # Another half step with kinetic energy
     Psi = np.matmul(UkinHalf, Psi)
     # Update time
     t = t + dt
  
  # Probability for remaining in the initial state
  Pinit = np.abs(np.trapz(np.conj(Psi)*Psi0, x))**2
  # Write result to screen
  print(f'omega: {omega:.4f}, Pinit: {Pinit:.4f}')
  Pvector[index] = Pinit
  index = index + 1

# Plot Pinit as a function of omega
plt.figure(1)
plt.clf()
plt.plot(omegaVector, Pvector, '-', linewidth = 2, color = 'black')
plt.xlabel(r'$\omega$', fontsize = 12)
plt.ylabel(r'$P_{init}$', fontsize = 12)
plt.grid()
plt.show()

File: MatlabCode/Chapter5_Time-dependent/Exercise_5_5_2.m
% This script calculates the probability to remain in the the 
% initial state for a particle in a harmonic osciallator which
% is shaken. It does so by first expressing the kinetic energy term
% via FFT and then approximating the second order Magnus propagator as 
% a split operator.
% 
% The harmonic oscillator potential is subject to a time-dependent 
% translation which shifts it in both directions before restoring it 
% in its original position. In addition to time, it takes omega as an 
% input; the duraton of the time-dependent shift is related to omega by 
% T = 2*pi/omega.
%
% Inputs
% L         - size of domain 
% N         - number of grid points (should be 2^n)
% Nt        - the number of time steps for each run
% k         - the strength of the potential
% omegaMin  - minimal omega value
% omegaMax  - maximal omega value
% domega    - step size in omega
% InitialState - the index of the initial state;  InitialState = 0 is 
% the ground state
%
% All inputs are hard coded initially

% Grid parameters
L = 10;
N = 512;              

% Strength of the harmonic oscillator
k = 1;

% Time parameters
Nstep = 500;
omegaMin = 0.05;
omegaMax = 1.5;
domega = 0.025;

% Index of initial state (ground state corresponds to zero)
InitialState = 0;

% Shape of the potential
Vpot = @(x) 0.5*k*x.^2;

% The displacement of the potential
Ftrans = @(t, omega) 8/3^(3/2)*sin(omega*t/2).^2.*sin(omega*t);

% Set up the grid
x = linspace(-L/2,L/2,N);
x = x.';                            % Transpose
h = L/(N-1);

% Set up kinetic energy matrix by means of the fast Fourier transform
k = 2*pi/(N*h)*[0:(N/2-1), (-N/2:-1)];          % Vector with k-values
% Fourier transform the identity matrix:
Tmat_FFT = fft(eye(N));
% Multiply by (ik)^2
Tmat_FFT = diag(-k.^2)*Tmat_FFT;
% Transform back to x-representation
Tmat_FFT = ifft(Tmat_FFT);
Tmat_FFT = -1/2*Tmat_FFT;            % Correct prefactor

% Diagonalize initial Hamiltonian
Hinit = Tmat_FFT + diag(Vpot(x));
[U E] = eig(Hinit);
E = diag(E);
[E Ind] = sort(E);
U = U(:, Ind)/sqrt(h);

% InitialState
Psi0 = U(:, InitialState+1);

% Loop over the omega values
omegaVector = omegaMin:domega:omegaMax;
Pvector = zeros(1, length(omegaVector));         % Allocate    
index = 1;
for omega = omegaVector
  % Duration
  tTotal = 2*pi/omega;
  % Time-step
  dt = tTotal/Nstep;
  % Propagator for half-step with kinetic energy
  UkinHalf = expm(-1i*Tmat_FFT*dt/2);
  t = 0;
  % Initial state
  Psi = Psi0;
  
  % Do the dynamics
  while t < tTotal
    % Half step with kinetic energy
    Psi = UkinHalf*Psi;  
    % Full step with potential energy
    Translation = Ftrans(t+dt/2, omega);
    UpotFull = diag(exp(-1i*Vpot(x-Translation)*dt));
    Psi = UpotFull*Psi;
    % Another half step with kinetic energy
    Psi = UkinHalf*Psi;
    % Update time
    t = t + dt;
  end
  % Probability for remaining in the initial state
  Pinit = abs(trapz(x,conj(Psi).*Psi0))^2;
  % Write result to screen
  disp(['omega: ', num2str(omega), ', Pinit: ', num2str(Pinit)])
  Pvector(index) = Pinit;
  index = index + 1;
end

% Plot Pinit as a function of omega
figure(1)
plot(omegaVector, Pvector, 'k-', 'linewidth', 2)
grid on
xlabel('\omega')
ylabel('P_{init}')
set(gca, 'fontsize', 15)